File: Vanilla/sources/zomboid/zombie/iso/Vector3.java
package zombie.iso;

import java.awt.Dimension;
import java.awt.Point;

public final class Vector3 implements Cloneable {
   public float x;
   public float y;
   public float z;

   public Vector3() {
      this.x = 0.0F;
      this.y = 0.0F;
      this.z = 0.0F;
   }

   public Vector3(Vector3 var1) {
      this.x = var1.x;
      this.y = var1.y;
      this.z = var1.z;
   }

   public Vector3(float var1, float var2, float var3) {
      this.x = var1;
      this.y = var2;
      this.z = var3;
   }

   public static Vector2 fromAwtPoint(Point var0) {
      return new Vector2((float)var0.x, (float)var0.y);
   }

   public static Vector2 fromLengthDirection(float var0, float var1) {
      Vector2 var2 = new Vector2();
      var2.setLengthAndDirection(var1, var0);
      return var2;
   }

   public static float dot(float var0, float var1, float var2, float var3) {
      return var0 * var2 + var1 * var3;
   }

   public void rotate(float var1) {
      double var2 = (double)this.x * Math.cos((double)var1) - (double)this.y * Math.sin((double)var1);
      double var4 = (double)this.x * Math.sin((double)var1) + (double)this.y * Math.cos((double)var1);
      this.x = (float)var2;
      this.y = (float)var4;
   }

   public void rotatey(float var1) {
      double var2 = (double)this.x * Math.cos((double)var1) - (double)this.z * Math.sin((double)var1);
      double var4 = (double)this.x * Math.sin((double)var1) + (double)this.z * Math.cos((double)var1);
      this.x = (float)var2;
      this.z = (float)var4;
   }

   public Vector2 add(Vector2 var1) {
      return new Vector2(this.x + var1.x, this.y + var1.y);
   }

   public Vector3 addToThis(Vector2 var1) {
      this.x += var1.x;
      this.y += var1.y;
      return this;
   }

   public Vector3 addToThis(Vector3 var1) {
      this.x += var1.x;
      this.y += var1.y;
      this.z += var1.z;
      return this;
   }

   public Vector3 div(float var1) {
      this.x /= var1;
      this.y /= var1;
      this.z /= var1;
      return this;
   }

   public Vector3 aimAt(Vector2 var1) {
      this.setLengthAndDirection(this.angleTo(var1), this.getLength());
      return this;
   }

   public float angleTo(Vector2 var1) {
      return (float)Math.atan2((double)(var1.y - this.y), (double)(var1.x - this.x));
   }

   public Vector3 clone() {
      return new Vector3(this);
   }

   public float distanceTo(Vector2 var1) {
      return (float)Math.sqrt(Math.pow((double)(var1.x - this.x), 2.0D) + Math.pow((double)(var1.y - this.y), 2.0D));
   }

   public float dot(Vector2 var1) {
      return this.x * var1.x + this.y * var1.y;
   }

   public float dot3d(Vector3 var1) {
      return this.x * var1.x + this.y * var1.y + this.z * var1.z;
   }

   public boolean equals(Object var1) {
      if (!(var1 instanceof Vector2)) {
         return false;
      } else {
         Vector2 var2 = (Vector2)var1;
         return var2.x == this.x && var2.y == this.y;
      }
   }

   public float getDirection() {
      return (float)Math.atan2((double)this.x, (double)this.y);
   }

   public Vector3 setDirection(float var1) {
      this.setLengthAndDirection(var1, this.getLength());
      return this;
   }

   public float getLength() {
      float var1 = this.getLengthSq();
      return (float)Math.sqrt((double)var1);
   }

   public float getLengthSq() {
      return this.x * this.x + this.y * this.y + this.z * this.z;
   }

   public Vector3 setLength(float var1) {
      this.normalize();
      this.x *= var1;
      this.y *= var1;
      this.z *= var1;
      return this;
   }

   public void normalize() {
      float var1 = this.getLength();
      if (var1 == 0.0F) {
         this.x = 0.0F;
         this.y = 0.0F;
         this.z = 0.0F;
      } else {
         this.x /= var1;
         this.y /= var1;
         this.z /= var1;
      }

      var1 = this.getLength();
   }

   public Vector3 set(Vector3 var1) {
      this.x = var1.x;
      this.y = var1.y;
      this.z = var1.z;
      return this;
   }

   public Vector3 set(float var1, float var2, float var3) {
      this.x = var1;
      this.y = var2;
      this.z = var3;
      return this;
   }

   public Vector3 setLengthAndDirection(float var1, float var2) {
      this.x = (float)(Math.cos((double)var1) * (double)var2);
      this.y = (float)(Math.sin((double)var1) * (double)var2);
      return this;
   }

   public Dimension toAwtDimension() {
      return new Dimension((int)this.x, (int)this.y);
   }

   public Point toAwtPoint() {
      return new Point((int)this.x, (int)this.y);
   }

   public String toString() {
      return String.format("Vector2 (X: %f, Y: %f) (L: %f, D:%f)", this.x, this.y, this.getLength(), this.getDirection());
   }

   public Vector3 sub(Vector3 var1, Vector3 var2) {
      return sub(this, var1, var2);
   }

   public static Vector3 sub(Vector3 var0, Vector3 var1, Vector3 var2) {
      var2.set(var0.x - var1.x, var0.y - var1.y, var0.z - var1.z);
      return var2;
   }
}

File: Vanilla/sources/zdoc/media/lua/shared/Library/Vector3.lua
---@class Vector3 : zombie.iso.Vector3
---@field public x float @The horizontal part of this vector
---@field public y float @The vertical part of this vector
---@field public z float
Vector3 = {}

---Make this vector identical to another vector
---@public
---@param other Vector3 @The Vector2 to copy
---@return Vector3
---@overload fun(x:float, y:float, z:float)
function Vector3:set(other) end

---Set the horizontal and vertical parts of this vector
---@public
---@param x float @The horizontal part
---@param y float @The vertical part
---@param z float
---@return Vector3
function Vector3:set(x, y, z) end

---Create a new vector from an AWT Point
---@public
---@param p Point @The java.awt.Point to convert
---@return JVector2 @A new Vector2 representing the Point
function Vector3:fromAwtPoint(p) end

---Returns a string representing this vector
---
---Overrides:
---
---toString in class java.lang.Object
---@public
---@return String @A String representing this vector
function Vector3:toString() end

---Convert this vector to an AWT Point
---@public
---@return Point @a java.awt.Point
function Vector3:toAwtPoint() end

---Set the direction of this vector, maintaining the length
---@public
---@param direction float @The new direction of this vector, in radians
---@return Vector3
function Vector3:setDirection(direction) end

---Add another vector to this one and return as a new vector
---@public
---@param other JVector2 @The other Vector2 to add to this one
---@return JVector2 @The result as new Vector2
function Vector3:add(other) end

---Set the direction of this vector to point to another vector, maintaining the length
---@public
---@param other JVector2 @The Vector2 to point this one at.
---@return Vector3
function Vector3:aimAt(other) end

---@public
---@return float
function Vector3:getLengthSq() end

---Set the length and direction of this vector
---@public
---@param direction float @The direction of this vector, in radians
---@param length float @The length of this vector
---@return Vector3
function Vector3:setLengthAndDirection(direction, length) end

---Create a new vector with a specified length and direction
---@public
---@param length float @The length of the new vector
---@param direction float @The direction of the new vector, in radians
---@return JVector2
function Vector3:fromLengthDirection(length, direction) end

---@public
---@param arg0 Vector3
---@return float
function Vector3:dot3d(arg0) end

---Add another vector to this one and store the result in this one
---@public
---@param other JVector2 @The other Vector2 to add to this one
---@return Vector3
---@overload fun(arg0:Vector3)
function Vector3:addToThis(other) end

---@public
---@param arg0 Vector3
---@return Vector3
function Vector3:addToThis(arg0) end

---Calculate the distance between this point and another
---@public
---@param other JVector2 @The second point as vector
---@return float @The distance between them
function Vector3:distanceTo(other) end

---Set the length of this vector, maintaining the direction
---@public
---@param length float @The length of this vector
---@return Vector3
function Vector3:setLength(length) end

---@public
---@param arg0 float
---@return void
function Vector3:rotatey(arg0) end

---Convert this vector to an AWT Dimension
---@public
---@return Dimension @a java.awt.Dimension
function Vector3:toAwtDimension() end

---Calculate the angle between this point and another
---@public
---@param other JVector2 @The second point as vector
---@return float @The angle between them, in radians
function Vector3:angleTo(other) end

---@public
---@param arg0 Vector3
---@param arg1 Vector3
---@return Vector3
---@overload fun(arg0:Vector3, arg1:Vector3, arg2:Vector3)
function Vector3:sub(arg0, arg1) end

---@public
---@param arg0 Vector3
---@param arg1 Vector3
---@param arg2 Vector3
---@return Vector3
function Vector3:sub(arg0, arg1, arg2) end

---Clone this vector
---
---Overrides:
---
---clone in class java.lang.Object
---@public
---@return Vector3
function Vector3:clone() end

---See if this vector is equal to another
---
---Overrides:
---
---equals in class java.lang.Object
---@public
---@param other Object @A Vector2 to compare this one to
---@return boolean @true if other is a Vector2 equal to this one
function Vector3:equals(other) end

---@public
---@return void
function Vector3:normalize() end

---@public
---@param other JVector2
---@return float
---@overload fun(x:float, y:float, tx:float, ty:float)
function Vector3:dot(other) end

---@public
---@param x float
---@param y float
---@param tx float
---@param ty float
---@return float
function Vector3:dot(x, y, tx, ty) end

---get the length of this vector
---@public
---@return float @The length of this vector
function Vector3:getLength() end

---@public
---@param arg0 float
---@return Vector3
function Vector3:div(arg0) end

---get the direction in which this vector is pointing
---
---Note: if the length of this vector is 0, then the direction will also be 0
---@public
---@return float @The direction in which this vector is pointing in radians
function Vector3:getDirection() end

---@public
---@param rad float
---@return void
function Vector3:rotate(rad) end
